Implement conditional, printing, and enabled/disabled breakpoints

ydeltastar opened the pull request:
- Closes: https://github.com/godotengine/godot-proposals/issues/909

Adds advanced breakpoints and related editor features to work with them.

![image](https://github.com/user-attachments/assets/7eec977d-158e-409b-a38d-5ddbda545d57)

### Features

- Conditional breakpoints using `Expression` evaluation.
- Print breakpoints using string interpolation. Supports member and stack variables `Node {name} : {event}`.
- Can disable suspending execution. Useful for print-only breakpoints.
- Enable/disable or edit through the improved Breakpoint tree on the debugger's bottom panel.
- Support for passing settings by command line. `-b res://script.gd:55|print=Print:%20{event}`

I didn't make relevant changes in this PR for the DAP support which is used by the `godot-vscode-plugin` but the groundwork is done should someone want to implement it.

ckaiser commented:
Tested, works as expected.

The conditions are pretty powerful, I'm a fan. It _does_ let you shoot yourself in the foot by calling functions with side effects, but I'm not opposed to that being a thing that's allowed.

One thing I could see being improved is the text for the "Print" placeholder, I had to resize the window to finish reading it and find out it's with `{var_name}`, it's a minor thing but the text could be condensed or the sizing adjusted.

Checking the validity of the expression within the dialog, if possible, would be great as well, so you can see if you've made a syntax error (I accidentally tried to use `||` instead of `or` for example).

Overall though this is great, very useful and I'd love to see it get merged. Did not look at the code or documentation, just functionality.

ydeltastar commented:
I modified the print hint text.

> Checking the validity of the expression within the dialog, if possible, would be great as well, so you can see if you've made a syntax error (I accidentally tried to use `||` instead of `or` for example).

`||` is a valid operator in GDScript.
I added an expression parsing error label but note that it can only catch basic syntax errors.

Calinou reviewed (commented):
Tested locally, it works as expected.

The idea of having non-suspending breakpoints is a great one; finally I may no longer have to rely on `print()` statement debugging so much :slightly_smiling_face: 

![image](https://github.com/user-attachments/assets/5ed934d9-150f-4de4-a2aa-d13d24ad8e06)

Some feedback:

- Using an unknown variable name (e.g. `counters >= 10` instead of `counter >= 10` in the above example) will fail silently. This should probably print an error when the line is first reached (but without breaking).
- Using an unknown print name (e.g. `Value of counter: {counters}` instead of `Value of counter: {counter}`) won't replace the placeholder and no error is printed. I'd expect it to print an error (and leave the placeholder as-is).
- Prints could make use of `print_line_rich()` instead of `print_line()` so you can use BBCode formatting within them. Right now, BBCode is printed as-is (`[color=red]{counter}[/color]` won't show up in red).
- When a breakpoint has **Suspend Execution** disabled, it should have a different icon to distinguish it from breakpoints that don't suspend execution. This could be an outlined red circle, for instance. For conditional non-suspending breakpoints, the `=` sign could be present in red in the middle.
- Since we have some space, we could rename **Suspend Execution** to **Suspend Execution if Condition is Met** just to be clear.
- Changes are occasionally lost after I close and reopen the Edit Breakpoint dialog. It's inconsistent and I can't determine a pattern as to why it's occurring.
- The error message that occurs when you use an invalid expression (I used `:` for my testing) isn't very clear. While the editor does point back to the script line where it occurred, it doesn't tell you that it's coming from an invalid conditional breakpoint. Ideally, it should do that, but also print the expression that was attempted to be parsed. (Printing the expression could also be done at the Expression class level.)

![image](https://github.com/user-attachments/assets/b321f669-d698-4a54-baa1-1be850a4cca8)

> (I accidentally tried to use `||` instead of `or` for example).

It's strange that this is invalid in Expression, considering it's valid in GDScript (even if the preferred syntax is `or` as per the [style guide](https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/gdscript_styleguide.html)).

ydeltastar commented:
> The idea of having non-suspending breakpoints is a great one; finally I may no longer have to rely on `print()` statement debugging so much 🙂

Also no need to search for lost `print()` calls with breakpoints since they all appear in the **Breakpoints** tree! Which reminds me that the tree should show the icons too so we know the type making it easier to find debug prints. I added it.

> * Using an unknown print name (e.g. `Value of counter: {counters}` instead of `Value of counter: {counter}`) won't replace the placeholder and no error is printed. I'd expect it to print an error (and leave the placeholder as-is).

That's tricky because the user know what variables to pass for placeholders in a script but the debugger don't. It has to pass all the local and member variables as input for the string interpolation so there will always be an `not all arguments converted during string formatting` error if you don't use all them.

As for the others cases, I improved the overall error reporting. They all should link to the source and line now.

> * Since we have some space, we could rename **Suspend Execution** to **Suspend Execution if Condition is Met** just to be clear.

It sounds verbose since there are no other cases the option is relevant. It makes more sense to include a description in the tooltip instead. I added a tooltip for **Enabled** too.

> * When a breakpoint has **Suspend Execution** disabled, it should have a different icon to distinguish it from breakpoints that don't suspend execution. This could be an outlined red circle, for instance. For conditional non-suspending breakpoints, the `=` sign could be present in red in the middle.

I added new icons for non-suspending breakpoints.

> * Changes are occasionally lost after I close and reopen the Edit Breakpoint dialog. It's inconsistent and I can't determine a pattern as to why it's occurring.

I can't reproduce it. I've been using this in my custom build for sometime now and never had this issue.

ckaiser commented:
Tested again, works well, great improvements!

I'm not 100% convinced by the icons, in Visual Studio there's a clear difference between what they call a "tracepoint" (a breakpoint that prints and continues executing), which shows up as a diamond shape, and any breakpoint that stops execution which is always a circle, I like this convention just because it's more visible at a glance.

IMO it should be something like:
 * A red circle is an enabled breakpoint that will suspend execution.
 * A red diamond is an enabled breakpoint that won't suspend execution.
 * A red circle with an icon in the middle (plus, minus, equal sign, whichever ends up looking best) is a conditional breakpoint that will suspend execution
 * A red diamond with an icon in the middle is a conditional breakpoint that will not suspend execution

Changing the color to gray or hollowing them out works well for marking them as disabled, and additionally we could have an icon for printing, and another for printing + conditional, but that might be slightly too noisy or hard to read.

An important feature that I believe should be added is a tooltip to the breakpoint that indicates what it does quickly, because right clicking all the way down the big menu to edit the breakpoint just to see what it does is cumbersome when editing code.

This PR vs Visual Studio:
![image](https://github.com/user-attachments/assets/7c5dc6cb-4fa1-4eb5-a3ed-f0d2a3f5a984)
(Another note is the icon for the conditional breakpoint appears to be blurry and the colors have very little contrast)

Of course, this is just my (very Visual Studio-y) opinion, whichever direction it ends up going this is gonna be a massive help scripting.

kitbdev commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -2542,6 +2609,10 @@ void ScriptTextEditor::_make_context_menu(bool p_selection, bool p_color, bool p
 	context_menu->add_shortcut(ED_GET_SHORTCUT("script_text_editor/toggle_comment"), EDIT_TOGGLE_COMMENT);
 	context_menu->add_shortcut(ED_GET_SHORTCUT("script_text_editor/toggle_bookmark"), BOOKMARK_TOGGLE);
 
+	if (has_breakpoint) {
+		context_menu->add_separator();
+		context_menu->add_shortcut(ED_GET_SHORTCUT("script_text_editor/edit_breakpoint"), EDIT_BREAKPOINT);
+	}
```
I think this should be at the top of the list. When right clicking on a breakpoint, editing it is the most relevant option.

ydeltastar replied:
I don't think it makes for a good user experience having an option move around in the same menu. The gutter should have its own context menu instead.

kitbdev commented on editor/plugins/script_text_editor.cpp:
```diff
@@ -638,13 +639,38 @@ bool ScriptTextEditor::is_unsaved() {
+			}
+			const Dictionary &dict = E;
+			const int &line = ((int)dict["line"]);
+			code_editor->get_text_editor()->set_line_as_breakpoint(line - 1, true, dict["enabled"], dict["suspend"], dict["condition"], dict["print"]);
+			EditorDebuggerNode::get_singleton()->set_breakpoint(script->get_path(), line, true, dict["enabled"], dict["suspend"], dict["condition"], dict["print"]);
```
Doesn't `set_line_as_breakpoint` emit a signal that connects the breakpoint to the debugger? Is this second line needed?

kitbdev reviewed (commented):
Seems to work well.
Some feedback:

The hover breakpoint color is too dark on dark theme:

![image](https://github.com/user-attachments/assets/89bb3a1b-dbad-48e3-98ea-5f9995cd0fa3)

The breakpoints panel in the debugger should use the same icons and colors as in the script.

ydeltastar commented on doc/classes/ScriptEditorBase.xml:
```diff
@@ -24,6 +24,13 @@
+			<param index="0" name="source" type="StringName" />
+			<param index="1" name="line" type="int" />
+			<description>
+				Emitted when the user requests an edition of a breakpoint's settings.
```
I don't want to, but since the class is exposed, all signals are also exposed with no way around it.

kitbdev replied:
Is it possible to move it to ScriptTextEditor? in ScriptEditor where it connects you should be able to only connect if it can be cast:

```cpp
	ScriptTextEditor *ste = Object::cast_to<ScriptTextEditor>(se);
	if (ste) {
		ste->connect("breakpoint_edit_requested", callable_mp(this, &ScriptEditor::_breakpoint_edit_request));
	}
```

kitbdev commented on scene/gui/code_edit.cpp:
```diff
@@ -1329,21 +1329,39 @@ bool CodeEdit::is_drawing_executing_lines_gutter() const {
 
 			Color use_color = theme_cache.breakpoint_color;
+			Dictionary dict = get_line_gutter_metadata(p_line, p_gutter);
+			if (!(bool)dict["bp_enabled"]) {
+				use_color = theme_cache.breakpoint_disabled_color;
+			} else if (!((String)dict["bp_print"]).is_empty() && !(bool)dict["bp_suspend"]) {
+				use_color = theme_cache.breakpoint_print_color;
+			}
```
```suggestion
			if (breakpointed && !(bool)dict["bp_enabled"]) {
				use_color = theme_cache.breakpoint_disabled_color;
			} else if (breakpointed && !((String)dict["bp_print"]).is_empty() && !(bool)dict["bp_suspend"]) {
				use_color = theme_cache.breakpoint_print_color;
			}
```
Since these should only be set if there actually is a breakpoint there.

kitbdev commented on scene/gui/code_edit.cpp:
```diff
@@ -1329,21 +1329,39 @@ bool CodeEdit::is_drawing_executing_lines_gutter() const {
 			icon_region.position += Point2(padding, padding);
 			icon_region.size -= Point2(padding, padding) * 2;
-			theme_cache.breakpoint_icon->draw_rect(get_canvas_item(), icon_region, false, use_color);
+			if (((String)dict["bp_condition"]).is_empty()) {
```
```suggestion
			if (!breakpointed) {
				theme_cache.breakpoint_icon->draw_rect(get_canvas_item(), icon_region, false, use_color);
			} else if (((String)dict["bp_condition"]).is_empty()) {
```
When hovering over a line without a breakpoint, It should use the regular breakpoint icon.

Also because on a regular CodeEdit that isn't part of the ScriptEditor, it shows a broken icon.

![image](https://github.com/user-attachments/assets/c8e6f9fd-a3b1-437e-83a3-237b09345a0c)

kitbdev reviewed (commented):
These are all of the combinations of icons:

|dark|light
|--|--|
|![image](https://github.com/user-attachments/assets/ecc08d42-14cf-4f88-8030-e643e740c02d) | ![image](https://github.com/user-attachments/assets/89ee9fa9-d1ba-45bd-a0cc-ef38a77a5a65) |

All combinations seem to work as expected functionally.

kitbdev reviewed (commented):
In the ScriptEditorDebugger, it is no longer possible to delete breakpoints when the script is not open in the ScriptEditor. When editing a breakpoint and the script is not open, the edit will be undone when the script is opened.
The "Delete All Breakpoints in: " option no longer shows the file path.

kitbdev commented on scene/gui/code_edit.cpp:
```diff
@@ -2902,7 +2951,12 @@ void CodeEdit::_bind_methods() {
 	BIND_THEME_ITEM(Theme::DATA_TYPE_ICON, CodeEdit, completion_color_bg);
 
 	BIND_THEME_ITEM(Theme::DATA_TYPE_COLOR, CodeEdit, breakpoint_color);
+	BIND_THEME_ITEM(Theme::DATA_TYPE_COLOR, CodeEdit, breakpoint_disabled_color);
+	BIND_THEME_ITEM(Theme::DATA_TYPE_COLOR, CodeEdit, breakpoint_print_color);
 	BIND_THEME_ITEM_CUSTOM(Theme::DATA_TYPE_ICON, CodeEdit, breakpoint_icon, "breakpoint");
+	BIND_THEME_ITEM_CUSTOM(Theme::DATA_TYPE_ICON, CodeEdit, breakpoint_no_suspend_icon, "breakpoint_no_suspend");
+	BIND_THEME_ITEM_CUSTOM(Theme::DATA_TYPE_ICON, CodeEdit, breakpoint_conditional_icon, "breakpoint_conditional");
+	BIND_THEME_ITEM_CUSTOM(Theme::DATA_TYPE_ICON, CodeEdit, breakpoint_conditional_no_suspend_icon, "breakpoint_conditional_no_suspend");
```
Hmm. I think having it also in ScriptTextEditor would be better than exposing these icons and needing to preserve this drawing behavior for compatibility.
I think we could use a callback to override just the breakpoint drawing part of it.

ydeltastar replied:
I can register only a single callback, so I can't override just the breakpoint drawing. I would need to make helper methods public in `CodeEdit` just for this callback, but these reviews have all been about avoiding exposing things.

I moved the gutter drawing to ScriptTextEditor, which helps with the other review case too.

kitbdev reviewed (commented):
Whenever I close and reopen a script, a new breakpoint is added at a line that isn't even in the script:

![image](https://github.com/user-attachments/assets/611bae3b-1ad5-4d29-bdfb-b99aae5611b7)

ydeltastar commented:
> Whenever I close and reopen a script, a new breakpoint is added at a line that isn't even in the script:

Fixed.

kitbdev commented:
When editing a breakpoint while the script is closed, it still gets reverted when opening the script. Also, if you restart the editor without opening the script, the edit also isn't saved.

ydeltastar committed:
Implement advanced breakpoints
